// @syncfusion/ej2-ng-base/src/form-base.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var FormBase = (function () {
        function FormBase() {
        }
        FormBase.prototype.propagateChange = function (_) { return; };
        FormBase.prototype.propagateTouch = function () { return; };
        FormBase.prototype.localChange = function (e) {
            var value = (e.checked === undefined ? e.value : e.checked);
            if (this.propagateChange !== undefined && value !== undefined) {
                this.propagateChange(value);
            }
        };
        FormBase.prototype.registerOnChange = function (registerFunction) {
            this.propagateChange = registerFunction;
        };
        FormBase.prototype.registerOnTouched = function (registerFunction) {
            this.propagateTouch = registerFunction;
        };
        FormBase.prototype.ngAfterViewInit = function () {
            var _this = this;
            setTimeout(function () {
                _this.appendTo(_this.element);
                var ele = _this.inputElement || _this.element;
                ele.addEventListener('focus', _this.ngOnFocus.bind(_this));
                ele.addEventListener('blur', _this.ngOnBlur.bind(_this));
            });
        };
        FormBase.prototype.setDisabledState = function (disabled) {
            this.enabled = !disabled;
        };
        FormBase.prototype.writeValue = function (value) {
            if (this.checked === undefined) {
                this.value = value;
            }
            else {
                if (typeof value === 'boolean') {
                    this.checked = value;
                }
                else {
                    this.checked = value === this.value;
                }
            }
            if (value === null) {
                return;
            }
        };
        FormBase.prototype.ngOnFocus = function (e) {
            this.focus.emit(e);
        };
        FormBase.prototype.ngOnBlur = function (e) {
            this.propagateTouch();
            this.blur.emit(e);
        };
        return FormBase;
    }());
    exports.FormBase = FormBase;
});
