// @syncfusion/ej2-ng-navigations/src/context-menu/contextmenu.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, ChangeDetectionStrategy } from '@angular/core';
import { ComponentBase, applyMixins, setValue } from '@syncfusion/ej2-ng-base';
import { ContextMenu } from '@syncfusion/ej2-navigations';
export var inputs = ['animationSettings', 'cssClass', 'enablePersistence', 'enableRtl', 'filter', 'items', 'locale', 'showItemOnClick', 'target'];
export var outputs = ['beforeClose', 'beforeItemRender', 'beforeOpen', 'onClose', 'onOpen', 'select'];
export var twoWays = [''];
/**
 * Represents the EJ2 Angular ContextMenu Component.
 * ```html
 * <div id='target'>Right click / Touch hold to open the ContextMenu</div>
 * <ej-contextmenu target='#target' [items]='menuItems'></ej-contextmenu>
 * ```
 */
var ContextMenuComponent = /** @class */ (function (_super) {
    __extends(ContextMenuComponent, _super);
    function ContextMenuComponent(ngEle, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.viewContainerRef = viewContainerRef;
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 2; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    ContextMenuComponent.prototype.ngOnInit = function () {
    };
    ContextMenuComponent.prototype.ngAfterViewInit = function () {
    };
    ContextMenuComponent.prototype.ngOnDestroy = function () {
    };
    ContextMenuComponent.prototype.ngAfterContentChecked = function () {
    };
    ContextMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ej-contextmenu',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    ContextMenuComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ViewContainerRef, },
    ]; };
    return ContextMenuComponent;
}(ContextMenu));
export { ContextMenuComponent };
applyMixins(ContextMenuComponent, [ComponentBase]);
