// @syncfusion/ej2-ng-navigations/src/toolbar/toolbar.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, ChangeDetectionStrategy, ContentChild } from '@angular/core';
import { ComponentBase, applyMixins, setValue } from '@syncfusion/ej2-ng-base';
import { Toolbar } from '@syncfusion/ej2-navigations';
import { ItemsDirective } from './items.directive';
export var inputs = ['enablePersistence', 'enableRtl', 'height', 'items', 'locale', 'overflowMode', 'width'];
export var outputs = ['beforeCreate', 'clicked', 'created', 'destroyed'];
export var twoWays = [''];
/**
 * Represents the Essential JS 2 Angular Toolbar Component.
 * ```html
 * <ej-toolbar></ej-toolbar>
 * ```
 */
var ToolbarComponent = /** @class */ (function (_super) {
    __extends(ToolbarComponent, _super);
    function ToolbarComponent(ngEle, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.viewContainerRef = viewContainerRef;
        _this.tags = ['items'];
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 2; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    ToolbarComponent.prototype.ngOnInit = function () {
    };
    ToolbarComponent.prototype.ngAfterViewInit = function () {
    };
    ToolbarComponent.prototype.ngOnDestroy = function () {
    };
    ToolbarComponent.prototype.ngAfterContentChecked = function () {
    };
    ToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ej-toolbar',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    queries: {
                        childItems: new ContentChild(ItemsDirective)
                    }
                },] },
    ];
    /** @nocollapse */
    ToolbarComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ViewContainerRef, },
    ]; };
    return ToolbarComponent;
}(Toolbar));
export { ToolbarComponent };
applyMixins(ToolbarComponent, [ComponentBase]);
