// @syncfusion/ej2-ng-navigations/src/accordion/accordion.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, ChangeDetectionStrategy, ContentChild } from '@angular/core';
import { ComponentBase, applyMixins, setValue } from '@syncfusion/ej2-ng-base';
import { Accordion } from '@syncfusion/ej2-navigations';
import { AccordionItemsDirective } from './items.directive';
export var inputs = ['animation', 'enablePersistence', 'enableRtl', 'expandMode', 'height', 'items', 'locale', 'width'];
export var outputs = ['clicked', 'created', 'destroyed', 'expanded', 'expanding'];
export var twoWays = [''];
/**
 * Represents the Essential JS 2 Angular Accordion Component.
 * ```html
 * <ej-accordion></ej-accordion>
 * ```
 */
var AccordionComponent = /** @class */ (function (_super) {
    __extends(AccordionComponent, _super);
    function AccordionComponent(ngEle, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.viewContainerRef = viewContainerRef;
        _this.tags = ['items'];
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 2; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    AccordionComponent.prototype.ngOnInit = function () {
    };
    AccordionComponent.prototype.ngAfterViewInit = function () {
    };
    AccordionComponent.prototype.ngOnDestroy = function () {
    };
    AccordionComponent.prototype.ngAfterContentChecked = function () {
    };
    AccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ej-accordion',
                    inputs: inputs,
                    outputs: outputs,
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    queries: {
                        childItems: new ContentChild(AccordionItemsDirective)
                    }
                },] },
    ];
    /** @nocollapse */
    AccordionComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ViewContainerRef, },
    ]; };
    return AccordionComponent;
}(Accordion));
export { AccordionComponent };
applyMixins(AccordionComponent, [ComponentBase]);
