// @syncfusion/ej2-ng-navigations/src/sidebar/sidebar.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, ViewContainerRef, ChangeDetectionStrategy } from '@angular/core';
import { ComponentBase, applyMixins, setValue } from '@syncfusion/ej2-ng-base';
import { Sidebar } from '@syncfusion/ej2-navigations';
export var inputs = ['animate', 'closeOnDocumentClick', 'contextTo', 'dockSize', 'enableDock', 'enableGestures', 'enablePersistence', 'enableRtl', 'height', 'locale', 'mediaQuery', 'position', 'showBackdrop', 'type', 'width', 'zIndex'];
export var outputs = ['created', 'change', 'close', 'destroyed', 'open'];
export var twoWays = [''];
/**
 * Represents the Essential JS 2 Angular Sidebar Component.
 * ```html
 * <ej-sidebar></ej-sidebar>
 * ```
 */
var SidebarComponent = /** @class */ (function (_super) {
    __extends(SidebarComponent, _super);
    function SidebarComponent(ngEle, viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.ngEle = ngEle;
        _this.viewContainerRef = viewContainerRef;
        _this.element = _this.ngEle.nativeElement;
        _this.injectedModules = _this.injectedModules || [];
        for (var i = 2; i < arguments.length; i++) {
            var injectedObj = arguments[i];
            if (injectedObj) {
                _this.injectedModules.push(injectedObj);
            }
        }
        _this.registerEvents(outputs);
        _this.addTwoWay.call(_this, twoWays);
        setValue('currentInstance', _this, _this.viewContainerRef);
        return _this;
    }
    SidebarComponent.prototype.ngOnInit = function () {
    };
    SidebarComponent.prototype.ngAfterViewInit = function () {
    };
    SidebarComponent.prototype.ngOnDestroy = function () {
    };
    SidebarComponent.prototype.ngAfterContentChecked = function () {
    };
    SidebarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ej-sidebar',
                    inputs: inputs,
                    outputs: outputs,
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    SidebarComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ViewContainerRef, },
    ]; };
    return SidebarComponent;
}(Sidebar));
export { SidebarComponent };
applyMixins(SidebarComponent, [ComponentBase]);
